pkg/webhook/clusterset/validatingWebhook.go

```diff
@@ -175,35 +172,91 @@ func (a *AdmissionHandler) validateCreateRequest(request *v1.AdmissionRequest) *
+//      2.3.2 if it is not claimed from clusterpool, check if it has permission to update clusterset.
+//
+// Permission Check:
+// 1. the user must have managedClusterSet/join <clusterSet-A/clusterSet-B> permission if update resource to change
+//         the managedClusterSet label from clusterSet-A to clusterSet-B.
+// 2. the user must have managedClusterSet/join <all> permission if update resource to add/remove the clusterSet label.
 func (a *AdmissionHandler) validateUpdateRequest(request *v1.AdmissionRequest) *v1.AdmissionResponse {
-	status := &v1.AdmissionResponse{
+	allowStatus := &v1.AdmissionResponse{
 		Allowed: true,
 	}
+	rejectStatus := &v1.AdmissionResponse{
+		Allowed: false,
+		Result: &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: "",
+		},
+	}
 
-	oldObj := unstructured.Unstructured{}
-	err := oldObj.UnmarshalJSON(request.OldObject.Raw)
+	//check if there is a update clusterset label request
+	isUpdateClusterSet, originalClusterSetName, currentClusterSetName, err := a.isUpdateClusterset(request)
 	if err != nil {
-		status.Allowed = false
-		status.Result = &metav1.Status{
-			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
-			Message: err.Error(),
+		rejectStatus.Result.Message = err.Error()
+		return rejectStatus
+	}
+	if !isUpdateClusterSet {
+		return allowStatus
+	}
+
+	switch request.Resource {
+	case managedClustersGVR:
+		managedCluster := &clusterv1.ManagedCluster{}
+		if err := json.Unmarshal(request.Object.Raw, managedCluster); err != nil {
+			rejectStatus.Result.Message = err.Error()
+			return rejectStatus
 		}
+
+		// Get managedcluster's clusterdeployment
+		// 1. No related clusterdeployment, check if it has permission to update this cluster.
+		// 2. The related clusterdeployment is created from clusterpool, do not allow to update this cluster.
+		// 3. The related clusterdeployment is not created from clusterpool, check if it has permission to update this cluster.
+		clusterDeployment, err := a.HiveClient.HiveV1().ClusterDeployments(managedCluster.Name).Get(context.TODO(), managedCluster.Name, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			rejectStatus.Result.Message = err.Error()
+			return rejectStatus
+		}
+
+		// the cluster is claimed from clusterpool, do not allow to update clustersetlabel.
+		if clusterDeployment != nil && clusterDeployment.Spec.ClusterPoolRef != nil {
+			rejectStatus.Result.Message = "Do not allow to update claimed cluster's clusterset."
+			return rejectStatus
+		}
+		break
+	case clusterPoolsGVR:
+		// do not allow to update clusterpool's clusterset
+		rejectStatus.Result.Message = "Do not allow to update clusterpool's clusterset label"
+		return rejectStatus
+	case clusterDeploymentsGVR:
+		// check if the user has permission to update the clusterset.
```

Comment:
why cluster deployment label is allowed to update if there is pool ref.

Reply:
In clusterdeployment controller, we need to add the set label to clusterdeployment follow pool's set.